src/commands/deployment/correlate.ts

```diff
@@ -0,0 +1,163 @@
+import {Logger, LogLevel} from '../../helpers/logger'
+import {retryRequest} from '../../helpers/retry'
+import {CI_PROVIDER_NAME, CI_ENV_VARS, GIT_REPOSITORY_URL, GIT_SHA} from '../../helpers/tags'
+import {getApiHostForSite, getRequestBuilder} from '../../helpers/utils'
+
+/**
+ * This command collects environment variables and git information to correlate commits from the
+ * source code repository to the configuration repo repository. This allows to connect pipelines triggering
+ * changes on the configuration repository to deployments from gitOps CD providers
+ */
+export class DeploymentCorrelateCommand extends Command {
+  public static paths = [['deployment', 'correlate']]
+
+  public static usage = Command.Usage({
+    category: 'CI Visibility',
+    description: 'Correlate gitOps CD deployments with CI pipelines.',
+    details: `
+      This command will correlate the pipeline with a gitOps CD deployment.\n
+      It does not work for every setup, see README for details.
+    `,
+    examples: [['Correlate an Argo CD deployment', 'datadog-ci deployment correlate --provider argocd']],
+  })
+
+  private cdProviderParam = Option.String('--provider')
+  private configurationRepo = Option.String('--config-repo')
+  private dryRun = Option.Boolean('--dry-run', false)
+
+  private config = {
+    apiKey: process.env.DATADOG_API_KEY || process.env.DD_API_KEY,
+  }
+
+  private logger: Logger = new Logger((s: string) => this.context.stdout.write(s), LogLevel.INFO)
+
+  public async execute() {
+    if (!this.config.apiKey) {
+      this.logger.error(
+        `Neither ${chalk.red.bold('DATADOG_API_KEY')} nor ${chalk.red.bold('DD_API_KEY')} is in your environment.`
+      )
+
+      return 1
+    }
+
+    if (!this.cdProviderParam) {
+      this.logger.error('Missing CD provider. It must be provided with --provider')
+
+      return 1
+    }
+    this.cdProviderParam = this.cdProviderParam.toLowerCase()
+
+    const tags = getCISpanTags() || {}
+
+    if (!this.validateTags(tags)) {
+      return 1
+    }
+
+    let envVars: Record<string, string> = {}
+    if (tags[CI_ENV_VARS]) {
+      envVars = JSON.parse(tags[CI_ENV_VARS])
+      delete tags[CI_ENV_VARS]
+    }
+    const ciEnv: Record<string, string> = {
+      ...tags,
+      ...envVars,
+    }
+
+    const git = simpleGit({
+      baseDir: process.cwd(),
+      binary: 'git',
+      maxConcurrentProcesses: 2, // max 2 git commands at the same time
+    })
+
+    const currentBranch = await gitCurrentBranch(git)
+    if (!currentBranch) {
+      this.logger.error('Could not get current branch')
+
+      return 1
+    }
+
+    let localCommitShas: string[]
+    if (this.configurationRepo) {
+      localCommitShas = await gitLocalCommitShas(git, currentBranch)
+    } else {
+      ;[this.configurationRepo, localCommitShas] = await Promise.all([
+        gitRepositoryURL(git),
+        gitLocalCommitShas(git, currentBranch),
```

Comment:
What happens if any of these fails?

Reply:
Good point, the API validation would fail and return an error message but I added a small check to avoid sending a request.